File: src/Core/RangeSet.fs
namespace Zeta.Core

/// RangeSet — the F# ferry (oracle #2 of TS/F#/C#/Rust) for the **RangeSet** primitive: a sparse
/// integer set in compact range notation (`"1-5,8,10-17"`). The TS reference
/// (`src/Core.TypeScript/range-set/`) authors the shared `golden-vectors.json`; this replays it:
/// `render (parse input)` equals the **canonical** form, and `contains` agrees. "The compilers
/// don't lie."
///
/// Canonical form (the cross-oracle byte-diff contract): ranges sorted, disjoint, NON-ADJACENT
/// (overlapping AND touching ranges coalesce, `1-3,4-6` → `1-6`), each emitted as `n` when
/// `lo = hi` else `lo-hi`, joined by `,` with no spaces; the empty set renders `""`. Non-negative
/// JS-safe integers (matching the shared int wire domain).
///
/// Result over throw: `parse` returns `Result<RangeSet, RangeSetFeedback>` (the rejection-vector
/// contract — a malformed token declines the SPECIFIC variant, matching across oracles).
module RangeSet =

    /// The shared JS-safe-integer ceiling (2^53 - 1) — the int wire domain (matches Bonsai).
    [<Literal>]
    let private MaxSafeInt = 9007199254740991L

    /// An inclusive integer range `(lo, hi)` with `lo <= hi`.
    type Range = int64 * int64

    /// A normalized set of ranges: sorted, disjoint, non-adjacent (the canonical invariant).
    type RangeSet = Range list

    /// The typed reasons `parse` declines — the shared cross-oracle rejection-vector contract.
    type RangeSetFeedback =
        /// A token (or sub-token) was not a non-negative safe integer.
        | NotInteger of token: string
        /// A range `lo-hi` had `lo > hi`.
        | InvertedRange of lo: int64 * hi: int64
        /// A structurally bad token (empty between commas, trailing comma, empty sub-token, too many dashes).
        | Malformed of token: string

    /// Parse a non-negative integer token strictly: digits only, within the safe-int range.
    let private parseNat (token: string) : int64 option =
        if token.Length > 0 && token |> Seq.forall (fun c -> c >= '0' && c <= '9') then
            match System.Int64.TryParse token with
            | true, n when n >= 0L && n <= MaxSafeInt -> Some n
            | _ -> None
        else
            None

    /// Normalize raw ranges into the canonical invariant: sort, then coalesce overlapping/adjacent.
    let private normalize (ranges: Range list) : RangeSet =
        ranges
        |> List.sortWith (fun (a0, a1) (b0, b1) -> if a0 <> b0 then compare a0 b0 else compare a1 b1)
        |> List.fold
            (fun acc (lo, hi) ->
                match acc with
                // coalesce when the next range overlaps OR touches the previous (lo <= phi + 1)
                | (plo, phi) :: rest when lo <= phi + 1L -> (plo, max phi hi) :: rest
                | _ -> (lo, hi) :: acc)
            []
        |> List.rev

    /// Parse compact range notation into a canonical `RangeSet`. Empty string → empty set.
    let parse (s: string) : Result<RangeSet, RangeSetFeedback> =
        let trimmed = s.Trim()

        if trimmed = "" then
            Ok []
        else
            let tokens = trimmed.Split(',')

            let rec loop i acc =
                if i >= Array.length tokens then
                    Ok(normalize (List.rev acc))
                else
                    let raw = tokens.[i]
                    let token = raw.Trim()

                    if token = "" then
                        Error(Malformed raw)
                    else
                        let parts = token.Split('-')

                        match parts.Length with
                        | 1 ->
                            match parseNat parts.[0] with
                            | Some n -> loop (i + 1) ((n, n) :: acc)
                            | None -> Error(NotInteger token)
                        | 2 ->
                            // an empty sub-token ("-3", "5-") is structurally missing, not a bad number
                            if parts.[0] = "" || parts.[1] = "" then
                                Error(Malformed token)
                            else
                                match parseNat parts.[0], parseNat parts.[1] with
                                | Some lo, Some hi -> if lo > hi then Error(InvertedRange(lo, hi)) else loop (i + 1) ((lo, hi) :: acc)
                                | None, _ -> Error(NotInteger parts.[0])
                                | _, None -> Error(NotInteger parts.[1])
                        | _ -> Error(Malformed token)

            loop 0 []

    /// Render a `RangeSet` to its canonical compact string.
    let render (rs: RangeSet) : string =
        rs
        |> List.map (fun (lo, hi) -> if lo = hi then string lo else sprintf "%d-%d" lo hi)
        |> String.concat ","

    /// Whether `n` is a member of the set (ranges are sorted, so the scan early-exits).
    let rec contains (rs: RangeSet) (n: int64) : bool =
        match rs with
        | [] -> false
        | (lo, hi) :: rest ->
            if n < lo then false
            elif n <= hi then true
            else contains rest n

    /// The union of two range sets, re-normalized to canonical form.
    let union (a: RangeSet) (b: RangeSet) : RangeSet = normalize (a @ b)

    /// Add a single integer to the set (returns a new canonical set).
    let add (rs: RangeSet) (n: int64) : RangeSet = normalize ((n, n) :: rs)

    /// The total count of integers covered by the set.
    let size (rs: RangeSet) : int64 = rs |> List.sumBy (fun (lo, hi) -> hi - lo + 1L)

File: src/Core.TypeScript/range-set/range-set.js
/**
 * range-set.ts — TS reference (oracle #1) for **RangeSet**: a sparse integer set in
 * compact range notation (`"1-5,8,10-17"`). The TS oracle authors the shared golden vectors
 * (`golden-vectors.json`); F#/C#/Rust ferry and replay them byte-for-byte: `render(parse(input))`
 * equals the **canonical** form, and `contains` agrees. "The compilers don't lie."
 *
 * Canonical form (the cross-oracle byte-diff contract): ranges **sorted**, **disjoint**, and
 * **non-adjacent** (auto-merged: overlapping AND touching ranges coalesce, `1-3,4-6 → 1-6`),
 * each emitted as `n` when start==end else `start-end`, joined by `,` with no spaces; the empty
 * set renders `""`. Non-negative JS-safe integers only.
 *
 * Result over throw: `parse` returns `Result<RangeSet, RangeSetFeedback>` (the rejection-vector
 * contract — a malformed token declines the SPECIFIC variant, matching across oracles).
 */
const ok = (value) => ({ ok: true, value });
const err = (error) => ({ ok: false, error });
const MAX_SAFE = 9007199254740991; // 2^53 - 1 — the shared safe-int ceiling (matches the other primitives)
/** Parse a non-negative integer token strictly: digits only, within the safe-int range. */
function parseNat(token) {
    if (!/^[0-9]+$/.test(token))
        return null;
    const n = Number(token);
    return Number.isSafeInteger(n) && n <= MAX_SAFE ? n : null;
}
/** Normalize raw ranges into the canonical invariant: sort, then coalesce overlapping/adjacent. */
function normalize(ranges) {
    if (ranges.length === 0)
        return [];
    const sorted = [...ranges].sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    const merged = [];
    for (const [lo, hi] of sorted) {
        const last = merged[merged.length - 1];
        // coalesce when the next range overlaps OR touches the previous (lo <= last.hi + 1)
        if (last && lo <= last[1] + 1) {
            last[1] = Math.max(last[1], hi);
        }
        else {
            merged.push([lo, hi]);
        }
    }
    return merged;
}
/** Parse compact range notation into a canonical {@link RangeSet}. Empty string → empty set. */
export function parse(s) {
    const trimmed = s.trim();
    if (trimmed === "")
        return ok([]);
    const ranges = [];
    for (const raw of trimmed.split(",")) {
        const token = raw.trim();
        if (token === "")
            return err({ kind: "Malformed", token: raw });
        const parts = token.split("-");
        if (parts.length === 1) {
            const n = parseNat(parts[0]);
            if (n === null)
                return err({ kind: "NotInteger", token });
            ranges.push([n, n]);
        }
        else if (parts.length === 2) {
            // an empty sub-token ("-3", "5-") is structurally missing, not a bad number
            if (parts[0] === "" || parts[1] === "")
                return err({ kind: "Malformed", token });
            const lo = parseNat(parts[0]);
            const hi = parseNat(parts[1]);
            if (lo === null)
                return err({ kind: "NotInteger", token: parts[0] });
            if (hi === null)
                return err({ kind: "NotInteger", token: parts[1] });
            if (lo > hi)
                return err({ kind: "InvertedRange", lo, hi });
            ranges.push([lo, hi]);
        }
        else {
            // more than one dash (e.g. "1-2-3") — structurally malformed
            return err({ kind: "Malformed", token });
        }
    }
    return ok(normalize(ranges));
}
/** Render a {@link RangeSet} to its canonical compact string. */
export function render(rs) {
    return rs.map(([lo, hi]) => (lo === hi ? `${lo}` : `${lo}-${hi}`)).join(",");
}
/** Whether `n` is a member of the set (ranges are sorted, so the scan early-exits). */
export function contains(rs, n) {
    for (const [lo, hi] of rs) {
        if (n < lo)
            return false;
        if (n <= hi)
            return true;
    }
    return false;
}
/** The union of two range sets, re-normalized to canonical form. */
export function union(a, b) {
    return normalize([...a, ...b]);
}
/** Add a single integer to the set (returns a new canonical set). */
export function add(rs, n) {
    return normalize([...rs, [n, n]]);
}
/** The total count of integers covered by the set. */
export function size(rs) {
    let total = 0;
    for (const [lo, hi] of rs)
        total += hi - lo + 1;
    return total;
}
